torch/onnx/_internal/fx/passes/type_promotion.py

```diff
@@ -180,6 +186,23 @@ def __eq__(self, __value: object) -> bool:
     def __hash__(self) -> int:
         return f"{type(self)}:{self.namespace}.{self.op_name}".__hash__()
 
+    def _consolidate_input_dtype(
+        self, computed_dtype: torch.dtype, result_dtype: torch.dtype
+    ) -> torch.dtype:
+        """
+        Although opmath is the right thing to do to retain on-par precision, it inserts
```

Comment:
Looks like it's about time to start the task on type promotion in torchlib... @justinchuby

Reply:
What do we need?

Reply:
It seems that if opmath is removed, we will have to upcast them inside onnx function if we don't have type prmotion in torchlib.

Reply:
For ops in onnx that does not have lower precision counterpart. e.g.
```
def celu(self, x: half):
    upcasted_x = op.Cast(x, float32)
    result = op.Celu(upcasted_x)
    return op.Cast(result, half)
```
I feel it should be easier for backend to reason if the upcasting is done within torchlib function, so there is a clear separation from model code casts, which are all outside of torchlib function.

Reply:
OK. So sounds like we can simply create torchlib variants for them, since the number is limited?

Reply:
I wonder if this approach solves the under case scenarios, cause if it does, we can get rid of nearest matching.

Reply:
https://github.com/microsoft/onnxscript/pull/1158